tests/test_advantage.py

```diff
@@ -0,0 +1,173 @@
+import unittest
+
+from requests.exceptions import HTTPError
+
+from webapp import advantage
+
+
+class Response:
+    """An HTTP response for testing."""
+
+    def __init__(self, status_code, content):
+        self.status_code = status_code
+        self._content = content
+
+    def raise_for_status(self):
+        if self.status_code != 200:
+            raise (HTTPError(response=self))
+
+    def json(self):
+        return self._content
+
+
+class Session:
+    """A request session for testing."""
+
+    def __init__(self, resp: Response):
+        self._resp = resp
+        self.request_kwargs = None
+
+    def request(self, **kwargs):
+        self.request_kwargs = kwargs
+        return self._resp
+
+
+def make_client(session: Session):
```

Comment:
The mock tooling you've written here is nice and elegant.

After trying a few different things, our most common attitude to mocking API calls like this is now to use real API URLs, and then use [VCR](https://pypi.org/project/vcrpy/1.5.2/) to save the responses, so future calls are mocked versions of the real API response from the time of the snapshot. We find that this gives us the benefit of quick mock responses while also giving us the opportunity to catch errors relating to subtle changes in the API responses.

Do you think that could work here? Or do you think there are clear reasons why your approach here is better?

Reply:
Never used vcrpy and it seems neat.
There are pros and cons of using a tool like that. On the bright side, once in a while you can actually execute a real API call, therefore checking that the client code works, and that the API is stable enough.
But some problems come to mind:
- how simple is to simulate error cases? how do you make the API server return an error, to check that the client deals with that correctly?
- how simple is to log in interactively in case the API is authenticated, for instance against USSO?
- did we review the code in that library? Can we trust that it does not have bugs that lead to always making HTTP requests, or to leaking secrets, in case auth is required?
- is it worth introducing a dependency when the actual mock code is more or less 30 lines, and the actual benefit is partially overruled by QA exercises, that needs to be done anyway?

Reply:
I can respond to some of those:

> how simple is to simulate error cases? how do you make the API server return an error, to check that the client deals with that correctly?

I don't think it provides anything along these lines, so if you're not able to come up with a real API call that produces the error you want to test, then you should probably bypass vcrpy in that case and manually create a test error case.

> how simple is to log in interactively in case the API is authenticated, for instance against USSO?

Not trivial. I think we've come up with workaround for this in the past... @jkfran might know more?

> did we review the code in that library? Can we trust that it does not have bugs that lead to always making HTTP requests, or to leaking secrets, in case auth is required?

No we haven't specifically reviewed the code (although we've read some of it as needed), just the same as for all the other dependencies of ubuntu.com. We know that it doesn't always make HTTP calls, as that would have been pretty obvious. It does nothing automatic to prevent leaking of secrets, but it lets you strip specific headers, so that's what we've done when necessary. But the onus is on the developer.

> is it worth introducing a dependency when the actual mock code is more or less 30 lines, and the actual benefit is partially overruled by QA exercises, that needs to be done anyway?

I think the dependency concern only has much weight if it's literally the only dependency needed for our test suite. Even if this is true now, I doubt it's going to be for much longer. The QA point is a general argument against testing, but the general wisdom is that the existence of QA does not excuse your tests from being thorough.

Reply:
Yeah I was not using the QA argument against testing in general, of course.
That's why this branch introduces unit tests for the ua contracts API client.
I am just not sure that vcrpy would make our tests much more thorough in this specific case, in which we want to check both happy and unhappy cases, and all requests are authenticated.
I am also not sure about the example you linked: when recording, and real HTTP requests are made, those test cases check that the target service is up and running and all pages return 200. But what is actually tested when not doing real requests?